python/ray/tune/suggest/search.py

```diff
@@ -0,0 +1,184 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from itertools import chain
+
+from ray.tune.suggest.variant_generator import generate_trials
+from ray.tune.experiment import Experiment
+from ray.tune.error import TuneError
+
+
+class SearchAlgorithm(object):
+    """Interface of an event handler API for hyperparameter search.
+
+    Unlike TrialSchedulers, SearchAlgorithms will not have the ability
+    to modify the execution (i.e., stop and pause trials).
+
+    Trials added manually (i.e., via the Client API) will also notify
+    this class upon new events, so custom search algorithms should
+    maintain a list of trials ID generated from this class.
+
+    See `ExistingVariants`.
+    """
+
+    def next_trial(self):
+        """Provides Trial objects to be queued into the TrialRunner.
+
+        Returns:
+            Trial|None: If SearchAlgorithm cannot be queried at
+                a certain time (i.e. due to constrained concurrency), this will
+                return None. Otherwise, return a trial.
+        """
+        raise NotImplementedError
+
+    def on_trial_result(self, trial_id, result):
+        """Called on each intermediate result returned by a trial.
+
+        This will only be called when the trial is in the RUNNING state.
+
+        Arguments:
+            trial_id: Identifier for the trial.
+        """
+        pass
+
+    def on_trial_complete(self,
+                          trial_id,
+                          result=None,
+                          error=False,
+                          early_terminated=False):
+        """Notification for the completion of trial.
+
+        Arguments:
+            trial_id: Identifier for the trial.
+            result (TrainingResult): Defaults to None. A TrainingResult will
+                be provided with this notification when the trial is in
+                the RUNNING state AND either completes naturally or
+                by manual termination.
+            error (bool): Defaults to False. True if the trial is in
+                the RUNNING state and errors.
+            early_terminated (bool): Defaults to False. True if the trial
+                is stopped while in PAUSED or PENDING state.
+        """
+        pass
+
+    def is_finished(self):
+        """Returns True if no trials left."""
+        raise NotImplementedError
+
+
+class ExistingVariants(SearchAlgorithm):
+    """Uses Tune's variant generation for resolving variables.
+
+    Custom search algorithms can extend this class easily by overriding the
+    `try_suggest` method, which will override conflicting fields from
```

Comment:
Can we combine try_suggest and next_trial?

Reply:
I think in the general case, making the user define the trial is a huge burden. 

Having a helper factory class or method (of which this is an attempt but perhaps the wrong incarnation) that takes care of that is very important.

Just look at: 

```python
yield Trial(
    trainable_name=spec["run"],
    config=new_config,
    trial_id=trial_id,
    local_dir=os.path.join(args.local_dir, output_path),
    experiment_tag=experiment_tag,
    resources=resources,
    stopping_criterion=spec.get("stop", {}),
    checkpoint_freq=args.checkpoint_freq,
    restore_path=spec.get("restore"),
    upload_dir=args.upload_dir,
    max_failures=args.max_failures)
```

Ideally, if a researcher wanted to try out a new modeling technique, ie SigOpt, returning a dict of the fields being searched over should be all that they need to do. For the current state of the PR, if I wanted to implement SigOpt, I could do so in 20 lines. Otherwise, I need to copy a bunch of boiler plate.

Reply:
I am fine with returning just the hyperparameter configurations here, no need to return an entire trial object.

Reply:
I think keeping only entry `try_suggest` would be more clear. Maybe `next_trial` should be a tool for researcher to use?